### PR TITLE
Check if item is playable before adding 'play from here' or 'play only this' to context menu

### PR DESCRIPTION
**Pull request by i96751414**
## Description

Kodi should check whether an item is playable before adding 'play from here' or 'play only this' to context menu.

## Motivation and Context

Currently Kodi does not check if an item is playable when adding context menu options. Thus, in items which are not 'folders' or script calls it will always add 'play from here' or 'play only this' to the context menu, even if they are explicitly marked as non playable.

## Types of change

- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [X] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [ ] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

File: xbmc/video/windows/GUIWindowVideoBase.cpp [hunk 1]
```diff
@@ -897,10 +897,12 @@ void CGUIWindowVideoBase::GetContextButtons(int itemNumber, CContextButtons &but
         buttons.Add(CONTEXT_BUTTON_PLAY_PARTYMODE, 15216); // Play in Partymode
       }
 
-      //if the item isn't a folder or script, is a member of a list rather than a single item
-      //and we're not on the last element of the list,
+      //if the item isn't a folder or script, is not explicitly marked as not playable,
+      //is a member of a list rather than a single item and we're not on the last element of the list,
       //then add add either 'play from here' or 'play only this' depending on default behaviour
```
Comment (howie-f) [inline xbmc/video/windows/GUIWindowVideoBase.cpp:902]: ... then add either ...

**Review by howie-f — CHANGES_REQUESTED** [review]

File: xbmc/video/windows/GUIWindowVideoBase.cpp [hunk 2]
```diff
@@ -897,10 +897,12 @@ void CGUIWindowVideoBase::GetContextButtons(int itemNumber, CContextButtons &but
...
       }
 
-      //if the item isn't a folder or script, is a member of a list rather than a single item
-      //and we're not on the last element of the list,
-      //then add add either 'play from here' or 'play only this' depending on default behaviour
-      if (!(item->m_bIsFolder || item->IsScript()) && m_vecItems->Size() > 1 && itemNumber < m_vecItems->Size()-1)
+      //if the item isn't a folder or script, is not explicitly marked as not playable,
+      //is a member of a list rather than a single item and we're not on the last element of the list,
+      //then add either 'play from here' or 'play only this' depending on default behaviour
+      if (!(item->m_bIsFolder || item->IsScript())
```
Comment (enen92) [inline xbmc/video/windows/GUIWindowVideoBase.cpp:903]: I'd suggest doing the opposite. Check if the property exists and is explicitly marked as false. You never know which type of items rely on this functionality. So, setting the isPlayable property to false will allow to disable the context item programmatically from addons while retaining the same behaviour as today for all other items.

**Review by enen92 — APPROVED** [review]

File: xbmc/video/windows/GUIWindowVideoBase.cpp [hunk 3]
```diff
@@ -897,10 +897,12 @@ void CGUIWindowVideoBase::GetContextButtons(int itemNumber, CContextButtons &but
...
+          && (!item->HasProperty("IsPlayable") || item->GetProperty("IsPlayable").asBoolean())
```
Comment (enen92) [inline xbmc/video/windows/GUIWindowVideoBase.cpp:904]: Condition is also wrong. HasProperty("IsPlayable") || something, will evaluate to true if the first condition matches. Means listitem.seProperty("isPlayable", "false") will enable the context item too

Comment (i96751414) [reply xbmc/video/windows/GUIWindowVideoBase.cpp:904]: Note that `HasProperty` is negated. At the moment I believe the context item is added only when IsPlayable is explicitly set to false.

Comment (enen92) [reply xbmc/video/windows/GUIWindowVideoBase.cpp:904]: You're right, missed the negation somehow :) think I need either to stop using the phone or buy one with a bigger screen.

Either way I think it's better to rewrite the second condition to the negation too. IsPlayable, afaik is just a property we can use from plugins to signal Kodi the item is playable. It is not used at a large extension (if at all) throughout the core .

Thanks for the contribution

**Comment by howie-f** [comment]
great, thanks. could you now please squash down to a single commit and then apply the clang-format-diff https://jenkins.kodi.tv/job/BuildMulti-PR/16079/artifact/PR17883.diff ?
feel free to ask if you don't know how to achieve this.

**Comment by i96751414** [comment]
> great, thanks. could you now please squash down to a single commit and then apply the clang-format-diff https://jenkins.kodi.tv/job/BuildMulti-PR/16079/artifact/PR17883.diff ?
> feel free to ask if you don't know how to achieve this.

Done :)

File: xbmc/video/windows/GUIWindowVideoBase.cpp [hunk 4]
```diff
@@ -897,10 +897,12 @@ void CGUIWindowVideoBase::GetContextButtons(int itemNumber, CContextButtons &but
...
-      //and we're not on the last element of the list,
-      //then add add either 'play from here' or 'play only this' depending on default behaviour
-      if (!(item->m_bIsFolder || item->IsScript()) && m_vecItems->Size() > 1 && itemNumber < m_vecItems->Size()-1)
+      // if the item isn't a folder or script, is not explicitly marked as not playable,
+      // is a member of a list rather than a single item and we're not on the last element of the list,
+      // then add either 'play from here' or 'play only this' depending on default behaviour
+      if (!(item->m_bIsFolder || item->IsScript()) &&
+          (!item->HasProperty("IsPlayable") || item->GetProperty("IsPlayable").asBoolean()) &&
+          m_vecItems->Size() > 1 && itemNumber < m_vecItems->Size() - 1)
       {
```
Comment (enen92) [inline xbmc/video/windows/GUIWindowVideoBase.cpp:906]: The likelihood of causing regressions here is still quite big here. Maybe do the following condition (although completely ugly...)?

```cpp
if ((!(item->m_bIsFolder || item->IsScript()) &&
          m_vecItems->Size() > 1 && itemNumber < m_vecItems->Size() - 1 &&
           (!item->HasProperty("IsPlayable") || item->GetProperty("IsPlayable").asBoolean())) || (item->HasProperty("IsPlayable") && !item->GetProperty("IsPlayable").asBoolean()))
```

Condition is ugly but at least there is a way to disable the context menu item from python without causing regressions

Comment (i96751414) [reply xbmc/video/windows/GUIWindowVideoBase.cpp:906]: `(item->HasProperty("IsPlayable") && !item->GetProperty("IsPlayable").asBoolean()))` is the same as `(!item->HasProperty("IsPlayable") || item->GetProperty("IsPlayable").asBoolean()))` negated... I think we don't want that

**Review by enen92 — CHANGES_REQUESTED** [review]
<s>In the current state this will certainly cause regressions because kodi doesn't place the `IsPlayable` property on listitems. AFAIK this is only a property used by plugins (see: https://github.com/xbmc/xbmc/blob/aa39210197323230c1c2ab086a6f2a18f39daf9a/xbmc/interfaces/builtins/PlayerBuiltins.cpp#L425) and is not used anywhere else throughout the codebase. So if you include the fact the property is there in the default condition kodi won't show the context menu item in a lot of places it used to</s>

edit: disregard, original condition is fine

File: xbmc/video/windows/GUIWindowVideoBase.cpp (hunk 4 above)
Comment (enen92) [reply xbmc/video/windows/GUIWindowVideoBase.cpp:906]: Yeah, keep forgetting this is to show the contextmenu item and not to hide it. Updated the example condition

Comment (enen92) [reply xbmc/video/windows/GUIWindowVideoBase.cpp:906]: disregard the comments, brainfart on my side. Original condition is okay.

**Review by enen92 — APPROVED** [review]

**Review by phunkyfish — APPROVED** [review]
👍

**Pull request merged**